src/cpg_flow/metamist.py

```diff
@@ -0,0 +1,617 @@
+        Parse str and create a AnalysisStatus object
+        """
+        return {v.value: v for v in AnalysisStatus}[name.lower()]
+
+
+class AnalysisType(Enum):
+    """
+    Corresponds to metamist Analysis types:
+    https://github.com/populationgenomics/sample-metadata/blob/dev/models/enums
+    /analysis.py#L4-L11
+
+    Re-defined in a separate module to decouple from the main metamist module,
+    so decorators can use `@stage(analysis_type=AnalysisType.QC)` without importing
+    the metamist package.
+    """
+
+    QC = "qc"
+    JOINT_CALLING = "joint-calling"
+    GVCF = "gvcf"
+    CRAM = "cram"
+    MITO_CRAM = "mito-cram"
+    CUSTOM = "custom"
+    ES_INDEX = "es-index"
+    COMBINER = "combiner"
+
+    @staticmethod
+    def parse(val: str) -> "AnalysisType":
+        """
+        Parse str and create a AnalysisStatus object
+        """
+        d = {v.value: v for v in AnalysisType}
+        if val not in d:
+            raise MetamistError(
+                f"Unrecognised analysis type {val}. Available: {list(d.keys())}"
+            )
+        return d[val.lower()]
+
+
+@dataclass
+class Analysis:
+    """
+    Metamist DB Analysis entry.
+
+    See the metamist package for more details:
+    https://github.com/populationgenomics/sample-metadata
+    """
+
+    id: int
+    type: AnalysisType
+    status: AnalysisStatus
+    sequencing_group_ids: set[str]
+    output: Path | None
+    meta: dict
+
+    @staticmethod
+    def parse(data: dict) -> "Analysis":
+        """
+        Parse data to create an Analysis object.
+        """
+        req_keys = ["id", "type", "status"]
+        if any(k not in data for k in req_keys):
+            for key in req_keys:
+                if key not in data:
+                    logging.error(f'"Analysis" data does not have {key}: {data}')
+            raise ValueError(f"Cannot parse metamist Sequence {data}")
+
+        output = data.get("output")
+        if output:
+            output = to_path(output)
+
+        a = Analysis(
+            id=int(data["id"]),
+            type=AnalysisType.parse(data["type"]),
+            status=AnalysisStatus.parse(data["status"]),
+            sequencing_group_ids=set([s["id"] for s in data["sequencingGroups"]]),
+            output=output,
+            meta=data.get("meta") or {},
+        )
+        return a
+
+
+class Metamist:
+    """
+    Communication with metamist.
+    """
+
+    def __init__(self) -> None:
+        self.default_dataset: str = get_config()["workflow"]["dataset"]
+        self.aapi = AnalysisApi()
+
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_exponential(multiplier=3, min=8, max=30),
+        retry=retry_if_exception_type(ServiceException),
+        reraise=True,
+    )
+    def make_retry_aapi_call(self, api_func: Callable, **kwargv: Any):
+        """
+        Make a generic API call to self.aapi with retries.
+        Retry only if ServiceException is thrown
+
+        TODO: How many retries?
```

Comment:
Maybe we can open a subsequent PR to handle all the TODO's in the codebase, before the MVP launch.